File: src/sandbox/python/runner.py
import importlib.util
import json
import time


def normalize(value):
    if isinstance(value, str):
        return value.strip()
    if value is None:
        return ""
    if isinstance(value, (dict, list)):
        return json.dumps(value, ensure_ascii=False, sort_keys=True)
    return str(value)


with open("tests.json", "r", encoding="utf-8") as test_file:
    payload = json.load(test_file)

tests = payload.get("tests", [])
entry_function = payload.get("entryFunction", "solve")

spec = importlib.util.spec_from_file_location("user_code", "main.py")
user_code = importlib.util.module_from_spec(spec)
spec.loader.exec_module(user_code)

start = time.time()
passed = 0
failed = 0
results = []

try:
    fn = getattr(user_code, entry_function)
except Exception:
    output = {
        "status": "ERROR",
        "passed": 0,
        "failed": len(tests),
        "total": len(tests),
        "executionMs": 0,
        "results": [],
        "error": f'Function "{entry_function}" not found in main.py',
    }
    print(json.dumps(output))
    raise SystemExit(0)

for test in tests:
    try:
        result = fn(test.get("input"))
        ok = normalize(result) == normalize(test.get("expected"))
        if ok:
            passed += 1
        else:
            failed += 1

        results.append(
            {
                "input": test.get("input"),
                "expected": test.get("expected"),
                "output": result,
                "passed": ok,
            }
        )
    except Exception as error:
        failed += 1
        results.append(
            {
                "input": test.get("input"),
                "expected": test.get("expected"),
                "passed": False,
                "error": str(error),
            }
        )

output = {
    "status": "PASSED" if failed == 0 else "FAILED",
    "passed": passed,
    "failed": failed,
    "total": len(tests),
    "executionMs": int((time.time() - start) * 1000),
    "results": results,
}

print(json.dumps(output))

File: src/sandbox/node/runner.js
const fs = require("fs");
const vm = require("vm");

const testsPayload = JSON.parse(fs.readFileSync("./tests.json", "utf-8"));
const userCode = fs.readFileSync("./user_code.js", "utf-8");
const entryFunction = testsPayload.entryFunction || "solve";
const tests = Array.isArray(testsPayload.tests) ? testsPayload.tests : [];

const normalize = (value) => {
  if (typeof value === "string") return value.trim();
  if (value === undefined || value === null) return "";
  if (typeof value === "object") return JSON.stringify(value);
  return String(value);
};

let solveFn = null;

try {
  const context = {
    module: { exports: {} },
    exports: {},
    require,
    console: {
      log: () => {},
      error: () => {},
      warn: () => {},
    },
    setTimeout,
    clearTimeout,
  };

  vm.createContext(context);
  vm.runInContext(userCode, context, { timeout: 1500 });

  if (typeof context.module.exports === "function") {
    solveFn = context.module.exports;
  } else if (
    context.module.exports &&
    typeof context.module.exports[entryFunction] === "function"
  ) {
    solveFn = context.module.exports[entryFunction];
  } else if (typeof context[entryFunction] === "function") {
    solveFn = context[entryFunction];
  }

  if (!solveFn) {
    throw new Error(
      `Function "${entryFunction}" not found. Export it or define it in global scope.`
    );
  }

  let passed = 0;
  let failed = 0;
  const results = [];
  const start = Date.now();

  for (const test of tests) {
    try {
      const output = solveFn(test.input);
      const ok = normalize(output) === normalize(test.expected);
      if (ok) passed += 1;
      if (!ok) failed += 1;

      results.push({
        input: test.input,
        expected: test.expected,
        output,
        passed: ok,
      });
    } catch (error) {
      failed += 1;
      results.push({
        input: test.input,
        expected: test.expected,
        passed: false,
        error: error instanceof Error ? error.message : String(error),
      });
    }
  }

  console.log(
    JSON.stringify({
      status: failed === 0 ? "PASSED" : "FAILED",
      passed,
      failed,
      total: tests.length,
      executionMs: Date.now() - start,
      results,
    })
  );
} catch (error) {
  console.log(
    JSON.stringify({
      status: "ERROR",
      passed: 0,
      failed: tests.length || 0,
      total: tests.length || 0,
      executionMs: 0,
      results: [],
      error: error instanceof Error ? error.message : String(error),
    })
  );
}
